cpp/src/arrow/python/deserialize.cc

```diff
@@ -108,153 +108,164 @@ Status DeserializeArray(int32_t index, PyObject* base, const SerializedPyObject&
-  OwnedRef result(CREATE_FN(stop_idx - start_idx));                                      \
-  const uint8_t* type_ids = data.raw_type_ids();                                         \
-  const int32_t* value_offsets = data.raw_value_offsets();                               \
-  for (int64_t i = start_idx; i < stop_idx; ++i) {                                       \
-    if (data.IsNull(i)) {                                                                \
-      Py_INCREF(Py_None);                                                                \
-      SET_ITEM_FN(result.obj(), i - start_idx, Py_None);                                 \
-    } else {                                                                             \
-      int64_t offset = value_offsets[i];                                                 \
-      uint8_t type = type_ids[i];                                                        \
-      PyObject* value;                                                                   \
-      RETURN_NOT_OK(GetValue(context, data, *data.UnsafeChild(type), offset, type, base, \
-                             blobs, &value));                                            \
-      SET_ITEM_FN(result.obj(), i - start_idx, value);                                   \
-    }                                                                                    \
-  }                                                                                      \
-  *out = result.detach();                                                                \
-  return Status::OK()
-
-Status DeserializeList(PyObject* context, const Array& array, int64_t start_idx,
-                       int64_t stop_idx, PyObject* base, const SerializedPyObject& blobs,
-                       PyObject** out) {
-  DESERIALIZE_SEQUENCE(PyList_New, PyList_SET_ITEM);
-}
-
-Status DeserializeTuple(PyObject* context, const Array& array, int64_t start_idx,
-                        int64_t stop_idx, PyObject* base, const SerializedPyObject& blobs,
-                        PyObject** out) {
-  DESERIALIZE_SEQUENCE(PyTuple_New, PyTuple_SET_ITEM);
+std::vector<int8_t> GetPythonTypes(const UnionArray& data) {
+  std::vector<int8_t> result;
+  auto type = data.type();
+  for (int i = 0; i < type->num_children(); ++i) {
+    result.push_back(std::stoi(type->child(i)->name()));
+  }
+  return result;
 }
 
-Status DeserializeSet(PyObject* context, const Array& array, int64_t start_idx,
-                      int64_t stop_idx, PyObject* base, const SerializedPyObject& blobs,
-                      PyObject** out) {
+template <typename CreateSequenceFn, typename SetItemFn>
+Status DeserializeSequence(PyObject* context, const Array& array, int64_t start_idx,
+                           int64_t stop_idx, PyObject* base,
+                           const SerializedPyObject& blobs,
+                           CreateSequenceFn create_sequence, SetItemFn set_item,
```

Comment:
AFAIK, the recommended way to pass the templated callables is to use "forwarding references" a.k.a. "universal references", in other words:
```
..., CreateSequenceFn&& create_sequence, SetItemFn&& set_item, ...
```

Reply:
fixed